controllers/etcdpeer_controller.go

```diff
@@ -307,10 +310,69 @@ func (r *EtcdPeerReconciler) maybeCreatePvc(ctx context.Context, peer *etcdv1alp
 	return true, nil
 }
 
+func hasPvcDeletionFinalizer(peer etcdv1alpha1.EtcdPeer) bool {
+	return sets.NewString(peer.ObjectMeta.Finalizers...).Has(pvcCleanupFinalizer)
+}
+
+// PeerPVCDeleter deletes the PVC for an EtcdPeer and removes the PVC deletion
+// finalizer.
+type PeerPVCDeleter struct {
+	log    logr.Logger
+	client client.Client
+	peer   *etcdv1alpha1.EtcdPeer
+}
+
+// Execute performs the deletiong and finalizer removal
+func (o *PeerPVCDeleter) Execute(ctx context.Context) error {
+	o.log.V(2).Info("Deleting PVC for peer prior to deletion")
+	expectedPvc := pvcForPeer(o.peer)
+	expectedPvcNamespacedName, err := client.ObjectKeyFromObject(expectedPvc)
+	if err != nil {
+		return fmt.Errorf("unable to get ObjectKey from PVC: %s", err)
+	}
+	var actualPvc corev1.PersistentVolumeClaim
+	err = o.client.Get(ctx, expectedPvcNamespacedName, &actualPvc)
+	switch {
+	case client.IgnoreNotFound(err) != nil:
+		return fmt.Errorf("failed to get PVC for deleted peer: %w", err)
+	case err == nil:
+		// PVC exists.
+		// Check whether it has already been deleted (probably by us).
+		// It won't actually be deleted until the garbage collector
+		// deletes the Pod which is using it.
+		if actualPvc.ObjectMeta.DeletionTimestamp.IsZero() {
+			o.log.V(2).Info("Deleting PVC for peer")
+			err := o.client.Delete(ctx, expectedPvc)
+			if err == nil {
+				o.log.V(2).Info("Deleted PVC for peer")
+				return nil
+			}
+			return fmt.Errorf("failed to delete PVC for peer: %w", err)
+		} else {
+			o.log.V(2).Info("PVC for peer has already been marked for deletion")
+		}
+	case err != nil:
```

Comment:
This is confusing - this case is *actually* only checking for the case where `IsNotFound == true`, because the first case check is checking if it *doesn't* equal NotFound.

Is the `switch` really required here? It reads easier (to me, at least) to have:

```
_, err := ...
if err == nil {
  // Delete PVC
}
if apierrors.IsNotFound(err) {
  // skip
} else if err != nil {
  return err
}
```

Or something similar? It just seems odd for the `if err != nil` case to actually only handle a very specific case of error

Reply:
I agree that last case is confusing. I couldn't remember where the the `IsNotFound` function is, so I tried to shoehorn in the `IgnoreNotFound` function which was to hand.

I've updated the switch statement to use `IsNotFound` and I prefer having it as a switch so that I know that only one of these cases can ever be executed.